rmk/src/direct_pin.rs

```diff
@@ -0,0 +1,319 @@
+    direct_pins: [[In; COL]; ROW],
+    /// Debouncer
+    debouncer: D,
+    /// Key state matrix
+    key_states: [[KeyState; COL]; ROW],
+    /// Start scanning
+    scan_start: Option<Instant>,
+    /// Pin active level
+    low_active: bool,
+}
+
+impl<
+        #[cfg(not(feature = "async_matrix"))] In: InputPin,
+        #[cfg(feature = "async_matrix")] In: Wait + InputPin,
+        D: DebouncerTrait,
+        const ROW: usize,
+        const COL: usize,
+    > DirectPinMatrix<In, D, ROW, COL>
+{
+    /// Create a matrix from input and output pins.
+    pub(crate) fn new(direct_pins: [[In; COL]; ROW], debouncer: D, low_active: bool) -> Self {
+        DirectPinMatrix {
+            direct_pins,
+            debouncer: debouncer,
+            key_states: [[KeyState::new(); COL]; ROW],
+            scan_start: None,
+            low_active,
+        }
+    }
+}
+
+impl<
+        #[cfg(not(feature = "async_matrix"))] In: InputPin,
+        #[cfg(feature = "async_matrix")] In: Wait + InputPin,
+        D: DebouncerTrait,
+        const ROW: usize,
+        const COL: usize,
+    > MatrixTrait for DirectPinMatrix<In, D, ROW, COL>
+{
+    const ROW: usize = ROW;
+    const COL: usize = COL;
+
+    #[cfg(feature = "async_matrix")]
+    async fn wait_for_key(&mut self) {
+        if let Some(start_time) = self.scan_start {
+            // If no key press over 1ms, stop scanning and wait for interupt
+            if start_time.elapsed().as_millis() <= 1 {
+                return;
+            } else {
+                self.scan_start = None;
+            }
+        }
+        Timer::after_micros(1).await;
+        info!("Waiting for active level");
+
+        // TODO: make it more efficient.
+        // `let mut futs: Vec<_, {ROW * COL}>` is invalid because of
+        // generic parameters may not be used in const operations.
+        // Maybe we can use nightly only feature `generic_const_exprs`
+        if self.low_active {
+            let mut futs: [Option<_>; ROW] = [const { None }; ROW];
+
+            for (row_idx, direct_pins_row) in self.direct_pins.iter_mut().enumerate() {
+                let mut futs_row: [Option<_>; COL] = [const { None }; COL];
+                for (col_idx, direct_pin) in direct_pins_row.iter_mut().enumerate() {
+                    futs_row[col_idx] = Some(direct_pin.wait_for_low());
+                }
+                futs[row_idx] = Some(select_array(futs_row.map(|option| option.unwrap())));
+            }
+            let _ = select_array(futs.map(|option| option.unwrap())).await;
+        } else {
+            let mut futs: [Option<_>; ROW] = [const { None }; ROW];
+
+            for (row_idx, direct_pins_row) in self.direct_pins.iter_mut().enumerate() {
+                let mut futs_row: [Option<_>; COL] = [const { None }; COL];
+                for (col_idx, direct_pin) in direct_pins_row.iter_mut().enumerate() {
+                    futs_row[col_idx] = Some(direct_pin.wait_for_high());
```

Comment:
Can be simplified with:

```rust
if self.low_active {
    futs_row[col_idx] = Some(direct_pin.wait_for_low());
} else {
    futs_row[col_idx] = Some(direct_pin.wait_for_high());
}
```

Reply:
`wait_for_low` and `wait_for_high` return different type of `Future`.
Simplifing like this will lead to [type inference error].(https://github.com/HaoboGu/rmk/pull/116/commits/0a6dd26847c0b6eafb708e4417e6a3e8fddeb2d6)

Reply:
I'm now using a 1-D heapless::Vec. The code is shorter, but simplifying it as you mentioned seems challenging.